#!/usr/bin/perl

# Script (c) Toni Mattila 2019 <toni+github@solu.fi>
# Included "MasterData.txt"s borrowed from http://dnagenics.com/tools/dna-kit-studio-v2-2/
# 23andme V3
# Usage: perl convert-evogenom-23andme.pl <SPORT_xxxxx_raakadata.txt >genome_xxxxx_V3.txt
# 23andme V4
# Usage: perl convert-evogenom-23andme.pl genome_V4_MasterData.txt <SPORT_xxxxx_raakadata.txt >genome_xxxxx_V4.txt
# 23andme V5
# Usage: perl convert-evogenom-23andme.pl genome_V5_MasterData.txt <SPORT_xxxxx_raakadata.txt >genome_xxxxx_V5.txt
# 23andme Combined V3-V5
# Usage: perl convert-evogenom-23andme.pl Full_MasterData.txt <SPORT_xxxxx_raakadata.txt >genome_xxxxx_V3_V4_V5.txt

my $master = shift || 'genome_V3_MasterData.txt';

open(F, "< $master");
while(my $line = <F>) {
  chomp($line);
  if($line!~/^#/) {
    my($id, $chrom, $pos, $gen) = split(/\t/, $line);
    $ref{"$chrom:$pos"} = $id;
  }
}

# skip first line of evogenom's file
my $firstline = <>;

print qq|# This data file generated by 23andMe at: Wed Jan 01 00:01:01 2018
#
# This file contains raw genotype data, including data that is not used in 23andMe reports.
# This data has undergone a general quality review however only a subset of markers have been 
# individually validated for accuracy. As such, this data is suitable only for research, 
# educational, and informational use and not for medical or other use.
# 
# Below is a text version of your data.  Fields are TAB-separated
# Each line corresponds to a single SNP.  For each SNP, we provide its identifier 
# (an rsid or an internal id), its location on the reference human genome, and the 
# genotype call oriented with respect to the plus strand on the human reference sequence.
# We are using reference human assembly build 37 (also known as Annotation Release 104).
# Note that it is possible that data downloaded at different times may be different due to ongoing 
# improvements in our ability to call genotypes. More information about these changes can be found at:
# https://you.23andme.com/p//tools/data/download/
# 
# More information on reference human assembly build 37 (aka Annotation Release 104):
# http://www.ncbi.nlm.nih.gov/mapview/map_search.cgi?taxid=9606
#
# rsid	chromosome	position	genotype
|;
my @result;

while(my $line = <>) {
  chomp($line);
  my($rscode, $chromesome, $position, $sampleid, $allele1, $allele2) = split(/\t/, $line);
  $allele1=~s{\"}{}g;
  $allele2=~s{\"}{}g;
  $rscode=~s{\"}{}g;
  $position=~s{\"}{}g;
  $chromesome=~s{\"}{}g;
  if($position > 0 && defined $ref{"$chromesome:$position"}) {
    push(@result, { 'chr' => $chromesome, 'pos' => $position, 'genotype' => $allele1.$allele2, 'rsid' => $ref{"$chromesome:$position"} });  
  }
}

my @sorted = sort { $a->{'chr'} cmp $b->{'chr'} or
                    $a->{'pos'} <=> $b->{'pos'}
                } @result;

foreach my $line (@sorted) {
  print join("\t", $line->{'rsid'}, $line->{'chr'}, $line->{'position'}, $line->{'genotype'}), "\n";
}
